go/decompress/decompress.go

```diff
@@ -0,0 +1,205 @@
+package decompress
+
+import (
+	"archive/tar"
+	"bufio"
+	"fmt"
+	"io"
+	"io/fs"
+	"log"
+	"os"
+	"strings"
+
+	"github.com/biogo/hts/bgzf"
+	gzip "github.com/klauspost/pgzip"
+)
+
+const EXPECTED_ANNOTATION_FILE_SUFFIX = "annotation.tsv.gz"
+
+type BystroReader interface {
+	ReadLines() ([]byte, error)
+	ReadLine() ([]byte, error)
+}
+
+type BzfBystroReader struct {
+	Reader *bgzf.Reader
+}
+
+type BufioBystroReader struct {
+	Reader *bufio.Reader
+}
+
+// Read a line up to the next newline character, and return the line excluding the newline character
+// Implementation follows the example in the bgzf package documentation: https://github.com/biogo/hts/blob/bb1e21d1bfc7f2b1e124ca0a1ed98493d191db78/bgzf/line_example_test.go#L70
+func _readLineBgzip(r *bgzf.Reader) ([]byte, error) {
+	var (
+		data []byte
+		b    byte
+		err  error
+	)
+	for {
+		b, err = r.ReadByte()
+		if err != nil {
+			break
+		}
+		if b == '\n' {
+			break
+		}
+		data = append(data, b)
+	}
+	return data, err
+}
+
+func ReadLineBgzip(r *bgzf.Reader) ([]byte, error) {
+	tx := r.Begin()
+	data, err := _readLineBgzip(r)
+	tx.End()
+	return data, err
+}
+
+func ReadLinesBgzip(r *bgzf.Reader) ([]byte, error) {
+	buf := make([]byte, 64*1024*8) // 8 bgzip blocks at a time
+
+	tx := r.Begin()
+	defer tx.End()
+
+	bytesRead, err := r.Read(buf)
+
+	if bytesRead == 0 {
+		return nil, err
+	}
+
+	if buf[bytesRead-1] != '\n' {
+		remainder, err := _readLineBgzip(r)
+		return append(buf[:bytesRead], remainder...), err
+	}
+	// last byte is newline
+	return buf[:bytesRead-1], err
+}
+
+func ReadLine(r *bufio.Reader) ([]byte, error) {
+	var (
+		data []byte
+		b    byte
+		err  error
+	)
+	for {
+		b, err = r.ReadByte()
+		if err != nil {
+			break
+		}
+		if b == '\n' {
+			break
+		}
+		data = append(data, b)
+	}
+	return data, err
+}
+
+func ReadLines(r *bufio.Reader) ([]byte, error) {
```

Comment:
Can you please add a test for this function? To be safe, testing with multiple threads seems like a good idea.